src/Processors/Transforms/DistinctSortedTransform.cpp

```diff
@@ -27,62 +27,92 @@ DistinctSortedTransform::DistinctSortedTransform(
         if (col && !isColumnConst(*col))
             column_positions.emplace_back(pos);
     }
+    /// pre-calculate DISTINCT column positions which form sort prefix of sort description
+    sort_prefix_positions.reserve(description.size());
+    for (const auto & column_sort_descr : description)
+    {
+        /// check if there is such column in header
+        if (!header.has(column_sort_descr.column_name))
+            break;
+
+        /// check if sorted column position matches any DISTINCT column
+        const auto pos = header.getPositionByName(column_sort_descr.column_name);
+        if (std::find(begin(column_positions), end(column_positions), pos) == column_positions.end())
+            break;
+
+        sort_prefix_positions.emplace_back(pos);
+    }
 }
 
 void DistinctSortedTransform::transform(Chunk & chunk)
 {
-        const ColumnRawPtrs column_ptrs(getKeyColumns(chunk));
-        if (column_ptrs.empty())
```

Comment:
I dropped this check because, AFAIU, it checks a consequence, not a root cause. `column_ptrs` will be empty if `column_positions` will be empty - which means either header doesn't contain columns or all columns are constant. So, if it's possible, then we should handle it earlier or differently.

@KochetovNicolai I'd be interested in your opinion. What is contract here? What make sense and what's not?

Reply:
Yes, the case with empty keys is unusual. If all keys are constants then there should be a single row in result and we can stop earlier.
Looks like previous check was just wrong.

Maybe we can process this case separately just in case.

Reply:
Will do in separate PR

Reply:
See #39688